keps/sig-node/4033-group-driver-detection-over-cri/README.md

```diff
@@ -0,0 +1,892 @@
+  runtime, making it unnecessary to specify cgroup driver in kubelet
+  configuration
+- maintain backward compatibility to run kubelet on top of an older container
+  runtime that doesn’t support this new feature
+
+### Non-Goals
+
+## Proposal
+
+### User Stories (Optional)
+
+#### Story 1
+
+As a cluster administrator, I would like to simplify my node configuration by
+configuring fields just once, and not needing to synchronize between multiple
+processes.
+
+#### Story 2
+
+As a novice Kubernetes user, I would like to easily be able to start pods with
+all runtimes, even if they have differing opinions on defaults.
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+### Risks and Mitigations
+
+Field adoption could be considered a risk, though the CRI implementations work
+closely with SIG Node and the feature will move along with CRI implementation
+adoption.
+
+## Design Details
+
+### CRI API
+
+Extend the CRI protocol to inform the kubelet which cgroup driver should be
+used.
+
+```diff
+ message RuntimeStatus {
+     // List of current observed runtime conditions.
+     repeated RuntimeCondition conditions = 1;
++    // Configuration settings of the runtime. This field contains global
++    // runtime configuration options that are not specific to runtime handlers.
++    RuntimeConfiguration configuration = 2;
+ }
+ 
++message RuntimeConfiguration {
++    // Configuration information for Linux-based runtimes
++    LinuxRuntimeConfiguration linux = 1;
++}
+ 
++message LinuxRuntimeConfiguration {
++    // Cgroup driver to use
++    CgroupDriver cgroup_driver = 1;
++}
++
++enum CgroupDriver {
++    CGGROUPFS = 0;
++    SYSTEMD = 1;
++}
+```
+
+The existing RuntimeStatus message (of the existing Status API endpoint) is
+used as this is being frequently queried by the kubelet, and is a place where
+the runtime tells the Kubelet about its state. The runtime will decide which
+CgroupDriver to choose based on existing methods: its own configuration.
+
+### Kubelet
+
+Kubelet will be modified to support the new field.
+
+If available the cgroup driver information received from the container runtime
+will take precedence over cgroupDriver setting from the kubelet config (or
+`--cgroup-driver` command line flag). If the runtime does not provide
+information about the cgroup driver, then kubelet will fall back to using its
+own configuration (`cgroupDriver` from kubeletConfig or the `--cgroup-driver`
+flag).
+
+Kubelet startup is modified so that connection to the CRI server (container
+runtime) is established and RuntimeStatus is queried before initializing the
+kubelet internal container-manager which is responsible for kubelet-side cgroup
+management.
```

Comment:
This is an important change of behavior, so if I have kubelet with hardcoded cgroup driver in version X and I upgrade to version X+1 I may have my kubelet running with a different configuration without noticing?

Reply:
I'd argue in this case N+1 would then actually fix your node. Logically there should be only one configuration point for this setting. Having it both in kubelet and container runtime is a constant source of problems as if your runtime has configured driver Y and kubelet has Z then your node is borken